Владимир Каткалов @katkalov

katkalof opened the pull request:
Запрошенное в задании поведение аналогично LinearLayout с layout_heigh=match_parent, искомый элемент match_parent в CustomLayout аналогичен элементу данного LL weight=n(у всех одно, если хотим несколько) и layout_width=0dp.

Рассмотрено большое количество различных случаев размещения CustomLayout и его сравнение с LL, в частности:
возможность  расположения нескольких элементов c width match_parent, расстояние делится поровну
отсутствие элементов с match_parent
расположения CustomLayout в ScrollView и HorizontalScrollView
установка margin и padding CustomLayout
различная высота вложенных View
и что-нибудь еще

Точно отсутствует:
margin у детей (интересно, как это происходит?)

![device-2016-07-27-072138](https://cloud.githubusercontent.com/assets/1605901/17163921/358c44ba-53cc-11e6-97bf-51f95dce0082.png)
![device-2016-07-27-072211](https://cloud.githubusercontent.com/assets/1605901/17163920/358b7abc-53cc-11e6-8cd2-1d60ebd0168b.png)
![device-2016-07-27-072220](https://cloud.githubusercontent.com/assets/1605901/17163922/358d72b8-53cc-11e6-8ac1-0bc51b9dc1ad.png)
![device-2016-07-27-072230](https://cloud.githubusercontent.com/assets/1605901/17163924/35900280-53cc-11e6-9648-39a3418dc2fa.png)
![device-2016-07-27-072240](https://cloud.githubusercontent.com/assets/1605901/17163923/358e0e12-53cc-11e6-81c8-5d7a3e05dbf6.png)

katkalof committed:
add https://github.com/yamblz-native/custom-viewgroup homework realisation and  check some case for it

bracadabra commented on app/src/main/java/ru/yandex/yamblz/ui/layouts/CustomLayout.java:
```diff
@@ -0,0 +1,97 @@
+        for (int i = 0; i < getChildCount(); ++i) {
+            View child = getChildAt(i);
+
+            if (child == null) {
```
А в какой ситуации такое может быть?

katkalof replied:
Я думаю, например, можно добавить программно вьюшку к лейауту зануленую, или удалить неправильно.
Читал исходники Linear'а, там есть подобная проверка - решил, что она не лишена смысла

bracadabra replied:
Если добавлять зануленную вьюху, то есть обычный null, то вылетет IllegaдArgumentException. Я посмотрел мельком LinealLayout, там вызвается getVirtualChildAt, судя по комменту, он используется в TableRow для каких-то выравниваний, то есть не подходит для нашего случая.

katkalof replied:
там под личиной метода getVirtualChildAt скрывается getChildAt(index)
по-этому я решил, что будет лучше перестраховаться и оставить данную проверку

bracadabra replied:
да, то есть использует реализацию по умолчанию, но вообще метод существует для TableRow, который видимо и может вернуть null.

katkalof replied:
осознал)
надо было внимательнее читать документацию)

katkalof replied:
fix

bracadabra commented on app/src/main/java/ru/yandex/yamblz/ui/layouts/CustomLayout.java:
```diff
@@ -0,0 +1,97 @@
+                    int measuredHeight = child.getMeasuredHeight();
+                    if (maxHeight < measuredHeight) {
+                        maxHeight = measuredHeight;
+                        changedHeightMeasureSpec = MeasureSpec.makeMeasureSpec(maxHeight, MeasureSpec.UNSPECIFIED);
```
UNSPECIFIED по идее не подразумевает никаких ограничений

katkalof replied:
Ну да, у меня и нет желания его ограничивать по высоте.
Пусть будет такая, сколько попросит самый высокий

bracadabra replied:
Это я к тому, что раз ограничений нет и maxHeight ни на что не влияет, то можно его и не передавать

katkalof replied:
смысл в том, что по maxHeight будет определена высота самого лейаута потом

bracadabra replied:
да, верно, значит мой коммент неактуален

bracadabra commented on app/src/main/java/ru/yandex/yamblz/ui/layouts/CustomLayout.java:
```diff
@@ -0,0 +1,97 @@
+        //Если осталось место или оно не ограничено, почему бы не показать matchParent
+        if ((totalWidth < layoutWidth) || widthMeasureSpec == MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED)) {
+            //В случае, если не знаем ширина родителя не остановленна,используем  вдвое увеличенную ширину не match_parent элементов
+            layoutWidth = layoutWidth != 0 ? layoutWidth : totalWidth * 2;
```
почему не утроенную ширину?

katkalof replied:
Можно и утроенную) 
Тут я размышлял о случае, когда вьюгрупп лежит в скроле
Получается надо как-то ограничивать match_parent чилда, либо вообще не давать, либо давать сколько-то

bracadabra replied:
Да, три тоже хорошее число) мне кажется, если в сумме все потомки превышают ширину экрана, то ширина match_parent вьюхи, должна быть просто 0, так как места для нее нет. Так получается более предсказуемо, чем умножать на 2, 3 и любое другое число.

katkalof replied:
=)
я делаю её для описанного тобой случая)
а потом решил придумать что-то поинтересней для скролвью, жалко стало вьюху)
согласен, что не супер вариант, но это же кастом :)

bracadabra replied:
ну да, теперь зачитаться можно)

katkalof replied:
сорян) у меня фантазия разыгралась тогда)

bracadabra commented on app/src/main/java/ru/yandex/yamblz/ui/layouts/CustomLayout.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Override
+    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
```
Нужно учитывать паддинги и при измерении

katkalof replied:
согласен, мой косяк
просто сравнив с linear'ом получил одинаковый результат(на скриншоте первом видно), и не стал задумываться
видимо тест не очень удачный
при первой возможности пофикшу

katkalof replied:
потыкал паддинги в методе измерения
ни к чему хорошему это не приводит)
наверно потому, что за меня это решает метод измерения суперкласса

bracadabra replied:
А можешь показать, где это супер класс делает?

katkalof replied:
Отказать, первое предположение было неверно
Осознал
Я вызываю для детей measureChild и он там вызывает getChildMeasureSpec и он думает, что я передаю родительские MeasureSpec поэтому хватает падинги родителя и смело из с их учетом расчитывает чилда. 
И если это так и происходит для первых детей, то для детей которые match_parent это не так. 

Это все будет работать до тех пор, пока match_parent в лейауте один. :)
то есть по сути задание выполнено)))
но надо бы исправить, чтобы для двух

katkalof replied:
Исправил) теперь учитываю паддинги нормально)
Сверху один match_parent, снизу 2 match_parent посередине
![device-2016-08-09-233722](https://cloud.githubusercontent.com/assets/1605901/17532653/55fb71da-5e8a-11e6-9aa0-7981e6668ef3.png)

bracadabra replied:
круто, я правда может не до конца понял твой замысел, но проблема была в том, что у match_parent вьюхи вызывается measure, которое ничего не знает о паддингах, поэтому возможна такая ситуация, что вьюха обрежется.

katkalof replied:
Примерно)
она, конечно, не обрежется - мы ей просто дадим сколько можем, как и подразумевается по заданию)
но дадим с учетом родительских паддингов, и потом все это будет работать, когда в лейаут напомним о их существовании
так что, получается для кейса, когда у нас есть один match_parent или вообще их нет в данном задании можно оставить так, как было. конечно теряется читаемость и осмысленность происходящего, но работать будет)

bracadabra replied:
Так view.measure не учитывает паддинги, паддинги учитывает measureChild, поэтому просто измерится по ширине и высоте родительской вьюхи без каких либо ограничений, а при отрисовке паддинг учитывается.

katkalof replied:
да я про прошлый случай и рассказывал, про measureChild. Типа почему это все работало раньше :) 
Пока тыкался сейчас, понял, что забыл проверять отсутствие match_parent элементов и делил на 0. Видимо, забыл добавить проверку, когда решил, что у меня может быть их больше одного. so sad 
пофиксил

bracadabra replied:
разобрались) в общем, это просто тонкий момент, который нужно в голове держать

katkalof replied:
понял. спасибо, было познавательно)
как-нибудь на досуге буду разбираться еще с мерджинами потомков, чтобы совсем проникнуться)

bracadabra replied:
без проблем

katkalof committed:
remove excess test

bracadabra commented:
По поводу марджинов у потомков. Для этого вьюха должна поддерживать MarginLayoutParams и соответственно учитывать их при измерении и расположении.

bracadabra commented on app/src/main/java/ru/yandex/yamblz/ui/layouts/CustomLayout.java:
```diff
@@ -0,0 +1,97 @@
+            if (child == null) {
+                continue;
+            }
+            if (child.getVisibility() == GONE) {
```
неплохо бы проверять скрыта ли вьюха и в onLayout

katkalof replied:
fix

katkalof committed:
add onlayout visibility==gone check

katkalof committed:
fix wrong work with 2 and more element and layout padding, add example to confirm
why - it was wrong using measureChild to element which width = match_parent
how - using measure for match_parent width children and consider layout padding in onMeasure

katkalof committed:
fix wrong work with 2 and more element and layout padding, add example to confirm
why - it was wrong using measureChild to element which width = match_parent
how - using measure for match_parent width children and consider layout padding in onMeasure

katkalof committed:
check divide by zero exception, when not match_parent children
